### DISCUSSION TITLE
Can I use a path parameter for a Rest API call or are only query parameters supported?

### DISCUSSION THREAD

**Discussion by thomasbergernz** [Budibase/budibase#7492] [Help] [answered]
The API endpoint I'm aiming to use is not supporting the use of a query param but a path param. 
While the datasource `https://api.something.org/api/compound-allocation-requests/{{id}}` works fine with **Binding name** `id` and default value `1234`, and so does the link from the table view (list all) to the individual record. However, the view always shows the default values of `1234`. 

When I utilise the _filtering_ option on the **Data Provider** like 
`id       Equals       Binding        {{ URL.id }}` the Repeater _Empty Text_ `{{ URL.id }} not found` is showing 
`7408 not found` because the loaded object is still `1234`

I'm using the self-hosted option deployed in a Kubernetes cluster.

**Comment by melohagan** [comment]
Hi @thomasbergernz 

I'm not quite sure what the problem is. It might help to see some screenshots, or if you could share your app export to support@budibase.com that would be helpful too!

**Comment by thomasbergernz** [comment]
Thanks for getting back to me @melohagan here is more context/background: 
I am evaluating the Rest API datasource option to build a CRUD app. 
I have not seen any templates or tutorials for this. 
I do have API endpoints for: 
GET /requests (how to support 'page number based' pagination?) [Docs cover cursor based pagination only](https://docs.budibase.com/docs/rest-pagination)
GET /requests/{id}
PUT /requests/{id}
POST /requests
DELETE /requests/{id}

My initial example is working to some extent, with the following setup:
`GET /requests/{id}` with `id` binding and no default value
![image](https://user-images.githubusercontent.com/10536616/188051274-f3a176ae-9dce-4474-8afd-09c80f5e92f4.png)
I've set the id manually, clicked 'Send', transformed the data (to parse a complex Json object), verified the Schema and clicked 'Save'
same for 
`GET /requests` to list all requests
![image](https://user-images.githubusercontent.com/10536616/188052260-6e740fbb-f085-4530-b23c-df0c11bcd872.png)
so the /requests 'Screen' has a 'Data Provider' and a table
![image](https://user-images.githubusercontent.com/10536616/188052809-cc7ade00-3ade-4186-b0ea-b4db2c004a85.png)
and the table configuration has 'Link table rows' activated, as well as the `id` column under 'Advanced' set.
![image](https://user-images.githubusercontent.com/10536616/188053283-581bb3cc-05cb-4841-bf61-79a21c0a1abe.png)
The rows now link to the detail screen and `id` gets appended to the path. 
Next, configure the detail screen 'Data' to use the `id` path param, by clicking the gear wheel icon next to it:
![image](https://user-images.githubusercontent.com/10536616/188053866-fb7651a9-7736-407e-8677-4313ff0621d6.png)
There I've set the 'Query Bindings' to `{{ URL.id }}`
![image](https://user-images.githubusercontent.com/10536616/188054058-98bebbb4-ffcd-414c-a39a-ee70f4a32fd4.png)
![image](https://user-images.githubusercontent.com/10536616/188054467-82f78eae-1cab-4b51-a0d7-fb201c42b037.png)
I checked that the Form has the correct schema:
![image](https://user-images.githubusercontent.com/10536616/188056062-303ba22a-dbfa-4f4f-ae94-1bc5983a2fd9.png)
When I add a 'New Field Group' component to the Form, and attempt to use the feature to [Generate Fields Automatically](https://docs.budibase.com/docs/forms#generating-fields-automatically) by clicking the 'Update form fields'
![image](https://user-images.githubusercontent.com/10536616/188056016-6b128dc5-8b0f-40c9-baa7-316e8ed16106.png)
I only get the `id` field generated. 

Is that a bug? What am I missing?

**Comment by melohagan** [comment, marked as answer]
Thanks for the extra info @thomasbergernz 

That has actually thrown me in the past as well! https://github.com/Budibase/budibase/issues/6238

I did add a note in the docs, but it is somewhat hidden/not obvious. It reads: 
> Similarly when using a [custom query](https://docs.budibase.com/docs/custom-queries) or [REST query](https://docs.budibase.com/docs/rest-queries) as the schema source, the form fields cannot be linked to the query schema.

Source: https://docs.budibase.com/docs/forms#form-schema

The reasoning behind this, from my understanding, is that typically form field to datasource mappings expect the ability to Read/Write.
For example given a Budibase internal DB datasource you could **read** values from the table schema within an Edit screen.
Then using the same form, you could use a **Save Row** action to **write** data to the same table. 

When you use a query (such as REST or custom SQL) you can only perform one CRUD operation at a time. Thus mapping the schema fields of a REST query into a form doesn't make a lot of sense. 
Sure you can read the fields from a GET request and use them in your form, but "saving" the form data would involve a POST or PUT and there's no guarantee that the fields from the GET response will match the body binding inputs of the PUT/POST.

Hopefully that made sense? The tricky part is trying to convey that in the docs, but I can put that existing sentence into a callout info card.

**Reply by melohagan** [reply]
I think the circumstance that bindings would be useful for REST queries would be as follows:

**REST Queries:**
- GET -- returns schema: { "ID", "Name", "Age", "DOB", "SSN" }
- PUT -- takes JSON payload with the following bindings:
```json
{
  "ID": {{ID}},
  "Name": {{Name}},
  "DOB": {{DOB}},
  "SSN": {{SSN}}
}
```

**Form**
Setup a form using the PUT query as it's schema. The names of the bindings match the field names in the GET schema, therefore you should be able to display the existing values with the GET query data provider. 
Then you can use the PUT query as your "save" action because the form fields match the bindings in the payload.
